modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C

```diff
@@ -0,0 +1,210 @@
+{
+}
+
+bool
+ApplyPenetrationConstraintLMMechanicalContact::shouldApply()
+{
+  const bool has_dofs = _current_node->n_dofs(_sys.number(), _var.number());
+  _dof_number = has_dofs ? _current_node->dof_number(_sys.number(), _var.number(), /*comp=*/0)
+                         : DofObject::invalid_id;
+  return has_dofs;
+}
+
+bool
+ApplyPenetrationConstraintLMMechanicalContact::overwriteSecondaryResidual()
+{
+  // If we didn't project the secondary node onto the primary surface, then gap has no meaning and
+  // we just want to drive our Lagrange Multiplier to zero. If we did, then our LM drives the
+  // residual if it is less than the gap (we're using min for our NCP function)
+  return (!_projection_successful) ||
+         (_u_secondary[_qp] < _node_to_integrated_gap.at(_current_node));
+}
+
+void
+ApplyPenetrationConstraintLMMechanicalContact::residualSetup()
+{
+  NodeFaceConstraint::residualSetup();
+  _node_to_integrated_gap.clear();
+}
+
+Real
+ApplyPenetrationConstraintLMMechanicalContact::computeQpSecondaryValue()
+{
+  return _u_secondary[_qp];
+}
+
+void
+ApplyPenetrationConstraintLMMechanicalContact::computeResidual()
+{
+  DenseVector<Number> & re = _assembly.residualBlock(_var.number());
+
+  _qp = 0;
+
+  _secondary_residual = re(0) = computeQpResidual(Moose::Secondary);
+  _secondary_residual_computed = true;
+}
+
+void
+ApplyPenetrationConstraintLMMechanicalContact::computeJacobian()
+{
+  _Kee.resize(1, 1);
+  _connected_dof_indices.clear();
+  _connected_dof_indices.push_back(_var.nodalDofIndex());
+
+  _qp = 0;
+
+  _Kee(0, 0) += computeQpJacobian(Moose::SecondarySecondary);
+}
+
+void
+ApplyPenetrationConstraintLMMechanicalContact::computeOffDiagJacobian(const unsigned int jvar_num)
+{
+  if (jvar_num == _var.number())
+  {
+    computeJacobian();
+    return;
+  }
+
+  const auto & jvar = getVariable(jvar_num);
+
+  _connected_dof_indices.clear();
+  _connected_dof_indices.push_back(jvar.nodalDofIndex());
+
+  _qp = 0;
+
+  _Kee.resize(1, 1);
+  _Kee(0, 0) += computeQpOffDiagJacobian(Moose::SecondarySecondary, jvar_num);
+
+  DenseMatrix<Number> & Ken =
+      _assembly.jacobianBlockNeighbor(Moose::ElementNeighbor, _var.number(), jvar_num);
+
+  for (_j = 0; _j < _phi_primary.size(); ++_j)
+    Ken(0, _j) += computeQpOffDiagJacobian(Moose::SecondaryPrimary, jvar_num);
+}
+
+Real
+ApplyPenetrationConstraintLMMechanicalContact::computeQpResidual(const Moose::ConstraintType)
+{
+  std::map<dof_id_type, PenetrationInfo *>::iterator found =
+      _penetration_locator._penetration_info.find(_current_node->id());
+  if (found != _penetration_locator._penetration_info.end())
+  {
+    PenetrationInfo * pinfo = found->second;
+    if (pinfo != NULL)
+    {
+      _projection_successful = true;
+      // We've stored the integrated gap in the residual vector
+      Real & integrated_gap = _node_to_integrated_gap[_current_node];
+      integrated_gap = _residual_copy(_dof_number) / _var.scalingFactor();
```

Comment:
Here, the scaling factor applied to the integrated gap corresponds to the primary variables? I.e. `disp_x` or `disp_y`?

Reply:
No it's the scaling factor for the LM variable. When we compute the weighted gap in our mortar constraint and save it into the residual vector, the scaling factor for the LM variable gets applied via multiplication. To undo that math, e.g. to make sure that our integrated gap here is truly representative of the gap value and not the gap value times some arbitrary factor, we divide by the scaling factor.